extensions/data-transfer/portability-data-transfer-google/src/main/java/org/datatransferproject/datatransfer/google/drive/DriveImporter.java

```diff
@@ -0,0 +1,148 @@
+      parentId = importSingleFolder(
+          jobId,
+          driveInterface,
+          "MigratedContent",
+          "root",
+          null);
+    } else {
+      DriveFolderMapping mapping = jobStore.findData(jobId, data.getId(), DriveFolderMapping.class);
+      checkNotNull(mapping, "No mapping found for %s", data.getId());
+      logger.info("Got parent id {} for old Id {} named: {}", parentId, data.getId(), data.getName());
+      parentId = mapping.getNewId();
+    }
+
+    // Uploads album metadata
+    if (data.getFolders() != null && data.getFolders().size() > 0) {
+      for (BlobbyStorageContainerResource folder : data.getFolders()) {
+        importSingleFolder(jobId, driveInterface, folder.getName(), folder.getId(), parentId);
+      }
+    }
+
+    // Uploads photos
+    if (data.getFiles() != null && data.getFiles().size() > 0) {
+      for (DigitalDocument file : data.getFiles()) {
+        importSingleFile(jobId, driveInterface, file, parentId);
+      }
+    }
+
+    return ImportResult.OK;
+  }
+
+  private String importSingleFolder(
+      UUID jobId,
+      Drive driveInterface,
+      String folderName,
+      String folderId,
+      String parentId) throws IOException {
+    File newFolder = new File()
+        .setName(folderName)
+        .setMimeType(DriveExporter.FOLDER_MIME_TYPE);
+    if (!Strings.isNullOrEmpty(parentId)) {
+      newFolder.setParents(ImmutableList.of(parentId));
+    }
+    File resultFolder = driveInterface.files().create(newFolder).execute();
+    DriveFolderMapping mapping = new DriveFolderMapping(folderId, resultFolder.getId());
+    jobStore.update(jobId, folderId, mapping);
+    return resultFolder.getId();
+  }
+
+  private void importSingleFile(
+      UUID jobId,
+      Drive driveInterface,
+      DigitalDocument file,
+      String parentId)
+      throws IOException {
+    InputStreamContent content = new InputStreamContent(
+        null,
+        jobStore.getStream(jobId, file.getCachedContentId()));
+    File driveFile = new File().setName(file.getName());
+    if (!Strings.isNullOrEmpty(parentId)) {
+      driveFile.setParents(ImmutableList.of(parentId));
+    }
+    if (!Strings.isNullOrEmpty(file.getDateModified())) {
+      driveFile.setModifiedTime(DateTime.parseRfc3339(file.getDateModified()));
+    }
+    if (!Strings.isNullOrEmpty(file.getOriginalEncodingFormat())
+        && file.getOriginalEncodingFormat().startsWith("application/vnd.google-apps.")) {
+      driveFile.setMimeType(file.getOriginalEncodingFormat());
+    }
+    driveInterface.files().create(
+        driveFile,
+        content
+    ).execute();
+  }
+
+  private synchronized Drive getDriveInterface(TokensAndUrlAuthData authData) {
+    if (driveInterface == null) {
+      driveInterface = makeDriveInterface(authData);
+    }
+
+    return driveInterface;
+  }
+
+  private synchronized Drive makeDriveInterface(TokensAndUrlAuthData authData) {
```

Comment:
are these the same in the exporter and importer? can they (along with a singleton) be moved to a common class or injected with a Provider perhaps if you want to lazy init?

Reply:
I made it static and combined the code.